functions.php

```diff
@@ -40,5 +40,5 @@
 
 // Include files.
 foreach ( $understrap_includes as $file ) {
-	require_once $understrap_inc_dir . $file;
+	require_once get_theme_file_path( $understrap_inc_dir . $file );
```

Comment:
Why would one want child themes to overload the parent theme's includes? It's the includes not the templates which are meant to be overloaded by child themes. This will most likely cause broken functionality due to missing functions. Each function is wrapped in `if ( ! function_exists( 'some_function_name' ) ) {}` and can easily be replaced with a function with the same name.

Reply:
The last version of Understrap committed to the WordPress repository used get_template_part() to allow overloading /inc/ files and sat there for two years telling people to do exactly this.

When we updated to 0.9.5 we broke a number of sites that relied on this behavior, so we restored it with the appropriate function.

I'm open to understanding your concern about missing functions, but if someone copies an /inc/ file to their child folder, as I understand it now, WP will load the file from the child theme's /inc/ and not finding any function by that name already, load the function. If they delete the function from the child theme copy, it will still be in the parent theme folder and thus, undeclared, be included from there.

Reply:
@danemorgan I'm confused.

`get_theme_file_path()` allows child themes to overload the parent theme. It first looks in the stylesheet directory for the file and only if it does not exist in the stylesheet directory it looks in the parent theme's directory aka the template directory.

`get_template_directory()` on the other hand never looks in the stylesheet directory unless there is no child theme and the stylesheet directory is equal to the template directory. Note the word template. For a child theme the parent theme is the template that's why you have to put the template in the child theme's style.css. Here the Understrap child telling WordPress that Understrap is its _template_:
https://github.com/understrap/understrap-child/blob/0bb163325b2d24f15ebfbdb3fa144b866bdbeb12/style.css#L7

Correct me if I'm wrong but using `get_theme_file_path()` allows child themes to do this:
```php
// inc/template-tags.php

function child_slug_something() {
  return 'No more template tags from parent will break my site';
}

function child_slug_something_else() {
  return 'Now I have to copy each function to my child theme';
}

function child_slug_final_thought() {
  return 'Why am I allowed to do this in the first place?';
}
```

Some Understrap breaking changes history:
On Sep 12, 2018 @ylkyrg introduced `locate_template()` in #805. I think the change to giving child themes the permission to overload include files was unintentional and no one else noticed. Of course problems popped up, see, e.g., #930. As a consequence, on Jul 30, 2019 @ylkyrg made a PR to switch back to `get_template_directory` (#1020) - right after the release of 0.9.4 which was the last one for a long time. At that time Understrap started to die if I remember correctly. The new owner approved #1020 on Oct 29, 2019 but did not manage to merge the PR until Dec 13, 2019. But there was no release anyway. Meantime, people started to adjust and obviously were told to overload include files. You breaking the breaking change, of course, caused problems again, which led to the switch to `get_theme_file_path()` which has the same functionality as `locate_template()`.

Reply:
I think this is an interesting discussion point- how should child theme functions OVERRIDE the parent theme? 

In previous versions we did it with a `function_exists()` wrapper around the functions themselves. That allowed functions to be overridden on a per-function basis. Then in the last "release" update we included the option to override the entire PHP file as if it were a template file. I'm tempted to vote towards returning to the previous method. 

The biggest reason being that any new functionality included in those other `/inc/` files will not work, meaning we'll have to be even MORE liberal with our use of `function_exists()` throughout the rest of the theme in case we add a new function to an `/inc/` file and then try to use it in a template.

I understand the debate and the fact that we've now publicly supported various methods, but I guess if I were casting a vote, I would return to the era of overriding single functions, not being able to override an entire `inc/*.php` file full of functionality and potentially missing out on future features. 

That being said- to play devil's advocate: To build a child theme is to adapt a parent theme *at a certain point in time* and you may not want new functionality always popping up on a site you very specifically put together with a particular set of tools...

Reply:
I think missing out on a new feature is less of a problem but breaking sites with new features is. It will become tremendously cumbersome to always take care of function which might or might not be there because a child theme may have overloaded include files.

> That being said- to play devil's advocate: To build a child theme is to adapt a parent theme at a certain point in time and you may not want new functionality always popping up on a site you very specifically put together with a particular set of tools...

I disagree on this one. That's the whole point of using a child theme! The parent theme authors fix bugs, add new features and so on. If I wanted to get stuck with a theme and do this all by myself, I'd just use the parent, give it another slug/name and start modifying that theme. Then there is no need for a child theme. But I am aware that Understrap is kinf of a hybrid between parent and starter theme.

I really do think that it would be best to think about a work around. E.g. checking for the existence of a file in the child theme and then load the parent theme's file on top of that. Haven't thought much about this, but maybe there is something we can do. At least for those doing a first install of Understrap we should not provide the possibility of overloading the include files. We could even add the option to check the child theme files for conflicts with `get_template_directory()`, no? The long run outcome should really be people replacing functions and/or template files with child themes and nothing more. Even if this means more problems and support requests because of breaking sites now. Imagine a future situation where each single change to Understrap may break a child theme. That's nothing I want to work on, do you?
You're planning a major release. It's the perfect time to do it now before more and more people jump on the overload bandwagon.

Reply:
I'm not insisting on the file override method. I restored it simply because when we pushed 0.9.5 to the WordPress repository we instantly started getting feedback that we had broken sites. I was not intending to advocate that position, only to explain the action and understand the ramifications more fully.

I actually thought of another potential problem, especially given the users involved. If someone overrides the file, then removes an entire function from the override, that function is gone, because the parent file isn't going to load, and that's got the potential to break the site too.

I'm fine with going back to stirctly using function overides, but we are going to have to manage a larger than insignificant WP repository users who are relying on that functionality.

Since our next version is going to be 1.0.0, I'd say we need to decide to do this now so it can go in there and not need another breaking release right away.

'm also in agreement with @IanDelMar's take that the purpose of a child theme is to allow continued updates and new features through the parent. If I want a feature-locked theme, I'm going to modify the core them, not use it as is with a child theme.

Reply:
Hi @danemorgan & @bacoords ! I had a look at a possible solution for this issue.

Basically, the following code first checks the child theme for the file. If it exists it loads the file. And then, no matter whether the file exist in the child theme or not all Understrap files get loaded. If a function has been defined in the child theme it won't be redefined because of all the `functions_exists()` checks and additionally because all Understrap files get loaded no function will be missing if a child theme overloaded a Understrap includes file in the past.

in functions.php
```diff
...

// Load Jetpack compatibility file if Jetpack is activiated.
if ( class_exists( 'Jetpack' ) ) {
	$understrap_includes[] = '/jetpack.php';
}

/*
 * In 2018 child themes were allowed to overload Understrap includes. This 
 * causes a lot of problems in the long run because Understrap is no longer in
 * control of its own includes.
 * Switching back to not allowing child theme to overload Understrap includes
 * introduces a breaking change for all child theme users. To mitigate the
 * effect we first load child theme files if they exist and load Understrap on
 * top. This ensures that each function of the child theme and each function
 * of Understrap gets loaded.
 */

// Include child theme files first.
+ if ( is_child_theme() ) {
+	foreach ( $understrap_includes as $file ) {
+		if ( file_exists( STYLESHEETPATH . $understrap_inc_dir . $file ) ) {
+			require_once get_stylesheet_directory() . $understrap_inc_dir . $file;
+		}
+	}
+}

// Load Understrap includes on top.
foreach ( $understrap_includes as $file ) {
-	require_once get_theme_file_path( $understrap_inc_dir . $file );
+	require_once get_template_directory() . $understrap_inc_dir . $file;
}
```

in inc/block-editor.php
```diff
...

if ( ! function_exists( 'understrap_generate_color_palette' ) ) {
	/**
	 * Checks for our JSON file of color values. If exists, creates a color palette array.
	 *
	 * @return array|bool
	 */
	function understrap_generate_color_palette() {
		$color_palette = array();

		// Grabs the autogenerated color palette that we're pulling from our compiled bootstrap stylesheets.
-		$color_palette_json = wp_remote_get( get_stylesheet_directory_uri() . '/inc/editor-color-palette.json' );
+		$color_palette_json = wp_remote_get( get_theme_file_uri( 'inc/editor-color-palette.json' ) );
		if ( ! is_wp_error( $color_palette_json ) ) {
			$color_palette_json = json_decode( $color_palette_json['body'], true );
			foreach ( $color_palette_json as $key => $value ) {
				$key             = ltrim( $key, '--' );
				$color_palette[] = array(
					'name'  => $key,
					'slug'  => $key,
					'color' => $value,
				);
			}
		}

		/**
		 * Filters the default bootstrap color palette so it can be overriden by child themes or plugins when we add theme support for editor-color-palette. This array can also be generated via gulp.
		 *
		 * @param array $color_palette An array of color options for the editor-color-palette setting.
		 */
		return apply_filters( 'understrap_theme_editor_color_palette', $color_palette );
	}
}

```

It worked fine in a first test on my local environment. Please feel free to test and tell me whether it worked for you as well.

If we implement this we could repeatedly tell people to check their files because we're going to deprecate the overloading sometime in the future. I think that's the long run goal. If one wants we could also add a database call and prohibit restrict existing users from overloading more files. I'm thinking about storing an option holding during the next update that holds all includes from the child theme and when loading the files only load those files from the child theme.

Assume a child theme comes with inc/extras.php and inc/hooks.php. On update we store the option understrap_overloads  with `array( 'extras.php' => true, 'hooks.php' => true )`. Then we can do the following in functions.php
```php
...

// Load Jetpack compatibility file if Jetpack is activiated.
if ( class_exists( 'Jetpack' ) ) {
	$understrap_includes[] = '/jetpack.php';
}

/*
 * In 2018 child themes were allowed to overload Understrap includes. This 
 * causes a lot of problems in the long run because Understrap is no longer in
 * control of its own includes.
 * Switching back to not allowing child theme to overload Understrap includes
 * introduces a breaking change for all child theme users. To mitigate the
 * effect we first load child theme files if they exist and load Understrap on
 * top. This ensures that each function of the child theme and each function
 * of Understrap gets loaded.
 */

// Include child theme files first.
 if ( is_child_theme() ) {
	$overloads = (array) get_option( 'understrap_overloads' );
	if ( ! empty( $overloads ) ) {
		foreach ( $understrap_includes as $file ) {
			if ( isset( $overloads[ $file ] ) ) {	
				if ( file_exists( STYLESHEETPATH . $understrap_inc_dir . $file ) ) {
					require_once get_stylesheet_directory() . $understrap_inc_dir . $file;
				}
			}
		}
	}
}
// Load Understrap includes on top.
foreach ( $understrap_includes as $file ) {
	require_once get_template_directory() . $understrap_inc_dir . $file;
}
```

I have no idea whether such a restriction that a child theme can not overload other files than it does at the time of the update would cause more problems than it solves.